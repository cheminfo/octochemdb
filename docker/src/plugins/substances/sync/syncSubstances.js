import firstSubstancesImport from './firstSubstancesImport.js';
import incrementalSubstancesImport from './incrementalSubstancesImport.js';
/**
 * @description start import of substances files and return the collection substances
 * @param {*} connection  - mongo connection
 */
export async function sync(connection) {
  await firstSubstancesImport(connection);
  await incrementalSubstancesImport(connection);
}
