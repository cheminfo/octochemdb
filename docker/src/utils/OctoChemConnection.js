import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
