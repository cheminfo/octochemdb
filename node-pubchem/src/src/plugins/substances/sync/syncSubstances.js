import firstSubstancesImport from './firstSubstancesImport.js';
import incrementalSubstancesImport from './incrementalSubstancesImport.js';

async function sync(connection) {
  await firstSubstancesImport(connection);
  await incrementalSubstancesImport(connection);
}

export default {
  sync,
};
