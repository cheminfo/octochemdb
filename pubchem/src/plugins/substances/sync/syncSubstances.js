import firstSubstancesImport from './firstSubstancesImport.js';
import incrementalSubstancesImport from './incrementalSubstancesImport.js';

export async function sync(connection) {
  await firstSubstancesImport(connection);
  await incrementalSubstancesImport(connection);
}
