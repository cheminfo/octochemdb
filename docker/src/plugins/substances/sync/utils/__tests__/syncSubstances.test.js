import { describe, it, expect } from 'vitest';

import { OctoChemConnection } from '../../../../../utils/OctoChemConnection.js';
import { sync } from '../../syncSubstances';

describe('syncSubstances', () => {
  it('syncSubstances First Importation', async () => {
    const connection = new OctoChemConnection();

    await sync(connection);
    const collection = await connection.getCollection('substances');
    const collectionEntry = await collection.find({ _id: 56427212 }).limit(1);
    const result = await collectionEntry.next();
    if (result?._seq) {
      delete result._seq;
    }
    expect(result).toMatchSnapshot();
    await connection.close();
  }, 30000);
  it('syncSubstances Incremental Importation', async () => {
    const connection = new OctoChemConnection();
    const collection = await connection.getCollection('substances');
    const collectionEntryIncremental = await collection
      .find({ _id: 475724937 })
      .limit(1);
    const resultIncremental = await collectionEntryIncremental.next();
    if (resultIncremental?._seq) {
      delete resultIncremental._seq;
    }
    expect(resultIncremental).toMatchSnapshot();
    await connection.close();
  });
}, 30000);
