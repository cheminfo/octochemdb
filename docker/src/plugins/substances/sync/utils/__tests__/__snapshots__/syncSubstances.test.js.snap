// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syncSubstances First Importation 1`] = `
Object {
  "_id": 56435271,
  "_seq": 6,
  "data": Object {
    "atom": null,
    "charge": 0,
    "cids": Array [
      15945538,
    ],
    "em": 358.16410520505997,
    "mf": "C18H22N4O4",
    "mw": 358.3923752735178,
    "nbFragments": 1,
    "ocl": Object {
      "acceptorCount": 8,
      "coordinates": "!BsMpiNNAefFsjGqcc?\`@@?iMgzsJH?Jh{Iz{~@Ha}b@JH@ha}b@JH?X\`B?g~H?Xc}?g|",
      "donorCount": 0,
      "idCode": "fegHPBHLtdBAJaJNrJZJJZZZEKIQILEEHymRuMUUT@E@@@",
      "index": Array [
        254,
        255,
        255,
        255,
        127,
        255,
        222,
        247,
        123,
        231,
        125,
        125,
        46,
        54,
        246,
        189,
        182,
        186,
        169,
        158,
        9,
        1,
        50,
        135,
        44,
        66,
        191,
        57,
        173,
        33,
        21,
        176,
        34,
        144,
        145,
        238,
        232,
        9,
        126,
        218,
        196,
        3,
        148,
        0,
        11,
        20,
        149,
        10,
        4,
        128,
        16,
        132,
        16,
        0,
        5,
        129,
        72,
        0,
        64,
        2,
        2,
        192,
        0,
        128,
      ],
      "logP": 1.6451000235974789,
      "logS": -2.627000069245696,
      "noStereoID": "fegHPBH\\\\tdeIJaMNrRJJJFZFEKIQI\\\\eEHytmMMUUT@E@@Gy\`@",
      "noStereoTautomerID": "fegHPBH\\\\tdBcVyMnrQJIJJJVEKIQIXIEHyuUUUUUT@E@@ZjQcFBXiafNXEbVMXgc}HXZccxX~F@",
      "polarSurfaceArea": 95.40000009536743,
      "rotatableBondCount": 5,
      "stereoCenterCount": 0,
    },
    "unsaturation": 10,
  },
  "naturalProduct": false,
}
`;

exports[`syncSubstances Incremental Importation 1`] = `
Object {
  "_id": 56435292,
  "_seq": 26,
  "data": Object {
    "atom": null,
    "charge": 0,
    "cids": Array [
      2335673,
    ],
    "em": 432.05467051102,
    "mf": "C20H17ClN2O5S",
    "mw": 432.8778667868712,
    "nbFragments": 1,
    "ocl": Object {
      "acceptorCount": 7,
      "coordinates": "!B~obSFN@ZkMa?oox@b@OMZ{tHbGpH_G~~B@N}BMHT_{z}B@Lw_wy?_\`Lw_wzMy@Lwtd|?TWoKV]|",
      "donorCount": 0,
      "idCode": "ek\`TJD@@jaIONi\`H\`\`HRYfy]~yW[eHXtlJjvNfhFBIjjffhd@@",
      "index": Array [
        254,
        253,
        127,
        255,
        95,
        255,
        221,
        215,
        223,
        146,
        125,
        125,
        45,
        21,
        242,
        61,
        180,
        152,
        137,
        138,
        169,
        57,
        54,
        130,
        60,
        98,
        174,
        11,
        41,
        96,
        141,
        162,
        2,
        40,
        152,
        171,
        238,
        9,
        102,
        66,
        204,
        32,
        20,
        8,
        2,
        60,
        132,
        17,
        0,
        194,
        48,
        212,
        28,
        16,
        17,
        132,
        0,
        64,
        64,
        1,
        2,
        0,
        0,
        168,
      ],
      "logP": 3.2176000624895096,
      "logS": -3.9499999545514584,
      "noStereoID": "ek\`TJD@@jmCOAej@cbHrRJJIZsISQIYQSJa\`RPjkXx\`fBZfjjejbP@C|p@",
      "noStereoTautomerID": "ek\`TJD@@jaIONi\`H\`\`HRYfy]~yW[eHXtlJjvNfhFBJjjffhd@CUBsCfOt\`sricxX~F@",
      "polarSurfaceArea": 108.29999876022339,
      "rotatableBondCount": 5,
      "stereoCenterCount": 0,
    },
    "unsaturation": 13,
  },
  "naturalProduct": false,
}
`;
